### PR TITLE
Remove support for reentrant application calls

### LINKED ISSUE #1152: Simplify system APIs
1. [ ] Unify the host-bindgen APIs used by “SimpleStateStorage” and “ViewStateStorage” (#1453, #1455)

    `contract_system_api.wit` and `service_system_api.wit` contain specific APIs only used by `SimpleStateStorage` and ultimately storing data in the map `ExecutionStateView::simple_users`.

    We should be able to remove these APIs and fix the modules `linera-sdk/src/{contract,service}/system_api.rs` so that the `SimpleStateStorage` feature is implemented on top of the same APIs as `ViewStateStorage` (e.g. `read_key` and `write_batch` from `view_system_api.wit`).

    This way, all user applications will use `ExecutionStateView::view_users` (to be renamed `ExecutionStateView::users` at the end).
2. [x] Remove locking primitive and disallow reentrancy by default (#1437, #1441)
3. [x] Make try_query_application_new synchronous (#1431)

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

We haven't yet ran into a scenario where reentrancy is required (at least that couldn't be solved using sessions). Support for reentrant calls is currently brittle, as it requires storing and loading the application state between calls, which might be unexpected to application developers. Adding proper support for reentrancy requires providing some new system APIs to provide fine-grain locking support. This also adds to the application development complexity and requires extra documentation and guidelines.

## Proposal

Remove support for reentrant applications for now. If there is ever the need for it, it can be added with a better implementation in the future.

## Test Plan

One of the unit tests includes an attempt to perform a reentrant call, which should fail.

## Release Plan

Changes the SDK API, so:
- Need to bump the major/minor version number in the next release of the crates.
- Need to update the developer manual.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

This is part of #1152, but does **not** close it.

### CODE REVIEW

File: linera-execution/src/runtime.rs [hunk 1]
```diff
@@ -39,7 +40,7 @@ pub struct SyncRuntimeInternal<const WRITABLE: bool> {
     execution_state_sender: ExecutionStateSender,
 
     /// The current stack of application descriptions.
-    applications: Vec<ApplicationStatus>,
+    applications: LinkedHashMap<UserApplicationId, ApplicationStatus>,
```
Comment (ma2bd) [inline linera-execution/src/runtime.rs:43]: This feels a bit overkill. I would just use keep the `Vec` and add a separate `HashSet<UserApplicationId>`.

File: linera-execution/tests/test_execution.rs [hunk 2]
```diff
@@ -62,22 +59,22 @@ async fn test_missing_bytecode_for_user_application() -> anyhow::Result<()> {
...
         result,
-        Err(ExecutionError::ApplicationBytecodeNotFound(desc)) if *desc == app_desc
+        Err(ExecutionError::ApplicationBytecodeNotFound(desc)) if &*desc == app_desc
     ));
     Ok(())
 }
 
 #[derive(Clone)]
-struct TestModule {
+struct TestModule<const CALLER: bool> {
```
Comment (ma2bd) [inline linera-execution/tests/test_execution.rs:68]: maybe `IS_CALLER?`

File: linera-sdk/src/lib.rs [hunk 3]
```diff
@@ -216,51 +216,39 @@ pub trait Contract: WithContractAbi + ContractAbi + Send + Sized {
     ) -> Result<SessionCallResult<Self::Message, Self::Response, Self::SessionState>, Self::Error>;
 
     /// Calls another application.
-    // TODO(#488): Currently, the application state is persisted before the call and restored after the call in
-    // order to allow reentrant calls to use the most up-to-date state.
-    async fn call_application<A: ContractAbi + Send>(
+    fn call_application<A: ContractAbi + Send>(
         &mut self,
```
Comment (ma2bd) [inline linera-sdk/src/lib.rs:220]: Nice! We have now a few refactoring options to discuss for a next PR (e.g. merge the function with `crate::contract::system_api::call_application` and/or make it a method of `*Context`).

File: linera-execution/src/runtime.rs (hunk 1 above)
Comment (jvff) [reply linera-execution/src/runtime.rs:43]: One disadvantage of having two types is that they must always be updated together. Forgetting to do so can lead to bugs :/

File: linera-execution/tests/test_execution.rs (hunk 2 above)
Comment (jvff) [reply linera-execution/tests/test_execution.rs]: Done.

File: linera-execution/src/runtime.rs (hunk 1 above)
Comment (ma2bd) [reply linera-execution/src/runtime.rs]: Thanks!

File: linera-execution/src/runtime.rs [hunk 4]
```diff
@@ -238,6 +241,42 @@ impl<const W: bool> SyncRuntimeInternal<W> {
...
+    ///
+    /// The current application is the last to be pushed to the `call_stack`.
+    ///
+    /// # Panics
+    ///
+    /// If the call stack is empty.
+    fn current_application(&mut self) -> &ApplicationStatus {
+        self.call_stack
+            .last()
+            .expect("Call stack is unexpectedly empty")
```
Comment (ma2bd) [inline linera-execution/src/runtime.rs:254]: Thanks for that one too!

**Review by ma2bd — APPROVED** [review]
Feels good to simplify the code!

**Pull request merged**